B2C Implementation Guide (#7310)
auth0/docs: articles/architecture-scenarios/b2c/authentication.md

---
title: Authentication

> description: Understand how authentication works in your CIAM implementation. 

Reviewer: I'm not sure a much better way to say this, maybe something like this:
```suggestion
description: Architect how you will make authentication work in your CIAM implementation.
```

toc: true
topics:
    - b2c
    - ciam
    - authentication
    - universal-login
contentType: concept
useCase:
    - authentication
---
# User Authentication

> In order to provide services to your users, you must be able to identify who those users are. This process is called authentication. You must provide your users a way to verify that they are who they say they are. There are a number of ways to do this (such as social media accounts, username/password, passwordless authentication). In most cases it is also recommended that you go beyond a first factor for authenticating the user and add a second factor as well.

Reviewer 2: ```suggestion
In order to provide services to your users, you must be able to identify who those users are. This process is called authentication. You must provide your users with a way to verify that they are who they say they are, and  there are a number of ways to do this: via social media accounts, username/password, and passwordless authentication to name but a few. In many cases it is also recommended that you go beyond a first factor for authenticating the user and add a second factor as well.
```

Reviewer 2: Multi-factor in B2C scenarios is not the norm, though there are cases where you would want to use it (such as in the financial sector, or some such).

Reviewer 3: My personal opinion here is that such as: or including: is better than a list followed by "to name but a few".

I'm also good with sticking to "most" unless it's a real sticking point - it's saying that we would _recommend_ that, not that we think most places currently use it. MFA _should_ probably be the norm for almost everything :grin:

Reviewer 2: To be honest I've no burning preference either way, so am happy to stick with the original :-)

> In order to provide services to your users, you must be able to identify who those users are. This process is called authentication. You must provide your users a way to verify that they are who they say they are. There are a number of ways to do this (such as social media accounts, username/password, passwordless authentication). In most cases it is also recommended that you go beyond a first factor for authenticating the user and add a second factor as well.

Reviewer: @yvonnewilson was concerned about this statement "they are who they say they are".  There is probably a better way to say this.  Perhaps "verify that they are allowed access to a particular identity in your system".

Reviewer 3: What's the concern exactly? I definitely think that "verify that they are allowed access to a particular identity in your system" is a mouthful of a phrase whose meaning isn't as clear.

It wouldn't be so bad to chop down the original sentence to "to verify who they are" or "to verify their identity" if either of those options are less objectionable?

> When designing my authentcation experience, how do I...?

Reviewer: ```suggestion
When designing my authentication experience, how do I...?
```

> * ...determine where users will enter their credentials?

Reviewer 2: Hmmm. I'm not sure bulleted ellipses like this are easy to follow. Just my personal opinion though - If it's a recognised Auth0 style then fine 👍

Reviewer 2: This is also inconsistent: articles/architecture-scenarios/b2c/branding.md for example uses the style as per the original B2C Welcome - which I also prefer. Either way, I think we should be consistent within the same set of documentation.

* ...keep user credentials safe?
* ...minimize the upkeep on my authentication system?
* ...provide password authentication for my users?
* ...integrate social logins for my users?
* ...prevent hackers from trying to log in as my users?
* ...implement authentication in different kinds of applications using Auth0?

> * ...make login easy for my users? I have customers from different language backgrounds?

Reviewer: ```suggestion
* ...make login easy for my users if they come from different language backgrounds?
```

* ...provide a good user experience as I migrate away from my legacy authentication system?

> The answers to these questions, and others, are detailed below in sections on [Universal Login](#universal-login), [Username Password Authentication](#username-password-authentication), [Anomaly Detection](#anomaly-detection), and [Application Integration](#application-integration).

Reviewer 2: The verbiage in the original B2C Welcome provided information directly and in a conversational fashion, without needing to seek out the answers. I feel that removing the verbiage loses something here.

Reviewer 2: The repetition also serves as reinforcement to the reader too.

Author: I will be going through all the docs and adding back in conversation tone and voice.

> The answers to these questions, and others, are detailed below in sections on [Universal Login](#universal-login), [Username Password Authentication](#username-password-authentication), [Anomaly Detection](#anomaly-detection), and [Application Integration](#application-integration).

Reviewer: Do we want to have a more clear distinction about which sections apply to which questions?

Reviewer 3: I think rather than add even more non-content than is already here (topic indices basically) I'd rather cut that paragraph to "The answers to these questions and more are detailed below". WDYT?

Reviewer 2: The intention behind this document is to explicitly guide the reader. The problem with "the answers to these questions and more are detailed below" is that the reader has to hunt through all the sections until they find the answer (to a specific the question). The verbiage in the original B2C Welcome was - rather than being specific content - actually designed to be narrative; introductory narrative if you like: the reader only has to go through one (or two) paragraphs until they find where to go next (to get the question answered). Does that make sense?

Author: Rewrite coming on all these to standardize the tone. I will make note of navigation to the pertinent paragraphs.

## Universal Login

Auth0's Universal Login makes authenticating users a short, three-step process (all of our [quickstarts](/quickstart) demonstrate this):

1. Determine how and when you want to [redirect from your application](#application-integration).
2. [Set up the appropriate branding and/or customized HTML](/universal-login) in your Auth0 configuration.
3. Set up your application to [receive and handle the response](#application-integration) from the Authorization Server.

::: note

> If you have more than one application, the best practice is to redirect to a centralized authorization server to authenticate the user (at Auth0, this means taking advantage of Universal Login). This comes with many security and user experience benefits, including SSO. See here for more information.

Reviewer 2: ```suggestion
If you have more than one application, the best practice is to redirect to a centralized authorization server to authenticate the user; at Auth0, this means taking advantage of Universal Login. This comes with many security and user experience benefits, including SSO. See here for more information.
```

> If you have more than one application, the best practice is to redirect to a centralized authorization server to authenticate the user (at Auth0, this means taking advantage of Universal Login). This comes with many security and user experience benefits, including SSO. See here for more information.

Reviewer: ```suggestion
If you have more than one application, the best practice is to redirect to a centralized authorization server to authenticate the user (at Auth0, this means taking advantage of Universal Login). This comes with many security and user experience benefits, including SSO. See the [Universal Login](/universal-login) documentation for more information.
```

:::

## Username Password Authentication

> Nearly every B2C application provides that ability for their customers to create a new set of credentials for this company. This is a common form of authentication that all users are familiar with.

Reviewer 2: ```suggestion
Nearly every B2C application provides the ability for their customers to create a new set of credentials. This is a common form of authentication that all users are familiar with.
```

> Username Password Authentication can come in multiple flavors at Auth0. If your application is a greenfield application with no existing user base, then a simple Auth0 OOTB database connection will give you everything you need to start authenticating your users. However if you have a legacy user store (such as your own database of users or an existing LDAP system) you have a couple of different options for migrating your users, see here for more information.

Reviewer 2: I typically favour a full explanation of an acronym the first time it is used. OOTB is an interesting one as it often gets written as OOB

Reviewer 2: Username Password authentication isn't just about migrating users either. This paragraph seems to be lacking the idea of also being able to use your existing identity store directly, which was conveyed in the original B2C Welcome document.

> Username Password Authentication can come in multiple flavors at Auth0. If your application is a greenfield application with no existing user base, then a simple Auth0 OOTB database connection will give you everything you need to start authenticating your users. However if you have a legacy user store (such as your own database of users or an existing LDAP system) you have a couple of different options for migrating your users, see here for more information.

Reviewer: Is it okay to assume that readers will know what OOTB means?

Author: removing OOTB and replacing it with "out-of-the-box" in all places

> Username Password Authentication can come in multiple flavors at Auth0. If your application is a greenfield application with no existing user base, then a simple Auth0 OOTB database connection will give you everything you need to start authenticating your users. However if you have a legacy user store (such as your own database of users or an existing LDAP system) you have a couple of different options for migrating your users, see here for more information.

Reviewer: ```suggestion
Username Password Authentication can come in multiple flavors at Auth0. If your application is a greenfield application with no existing user base, then a simple Auth0 OOTB database connection will give you everything you need to start authenticating your users. However if you have a legacy user store (such as your own database of users or an existing LDAP system) you have a couple of different options for migrating your users, see [User Migration](/user-provisioning#user-migration) for more information.
```

Reviewer: That said => @yvonnewilson made a great point the other day, that it may not be good to hyperlink out of the docs and instead have a section for references for more information in a particular section at the bottom.  See reference here: https://rylanschaeffer.github.io/content/books/shallows_carr_nicholas.html

Reviewer: @tamigoodall ^

Reviewer 3: It's pretty standard for us to link to other sections in our docs - I think that there's no need to try to change that on this one link, but it's definitely a valid question to bring up in general - whether we should, as we work on our docs IA, do less cross referencing and links within documents. I for one as a user find in-text links to be far more valuable than footnotes, but that's anecdotal.

In any case, I'd say let's fix this link as you've recommended and table the larger conversation. @tamigoodall WDYT?

Author: I agree. This is a larger question. We have already made some changes about how we reference links (inline always, inline sometimes, at the bottom under "Keep reading," but we haven't settled on a consistent approach. As we do more work on IA in the next quarter, we will make these all consistent with a chosen style and criteria for when we provide links inline and when we add links at the bottom.

> However you end up provisioning the users for your database connection, the authentication of those users is quite similar. It requires you to present the users with a form to enter their username and password. As mentioned in the universal login section, the simplest and safest way to authenticate users with a username and password is to redirect them to a centralized login page and collect their username and password there. This allows your authorization server to determine whether they have already authenticated and skip the login form entirely when it is not needed. See here for more information.

Reviewer 2: ```suggestion
However you end up provisioning the users for your database connection, the authentication of those users is similar: it requires you to present the users with a form to enter their username and password. As mentioned in the universal login section, the simplest and safest way to authenticate users with a username and password is to redirect them to a centralized login page where username and password will be collected. This allows your authorization server to determine whether a user has already authenticated, and skip the login form entirely when not required. See here for more information.
```

> However you end up provisioning the users for your database connection, the authentication of those users is quite similar. It requires you to present the users with a form to enter their username and password. As mentioned in the universal login section, the simplest and safest way to authenticate users with a username and password is to redirect them to a centralized login page and collect their username and password there. This allows your authorization server to determine whether they have already authenticated and skip the login form entirely when it is not needed. See here for more information.

Reviewer: ```suggestion
However you end up provisioning the users for your database connection, the authentication of those users is quite similar. It requires you to present the users with a form to enter their username and password. As mentioned in the universal login section, the simplest and safest way to authenticate users with a username and password is to redirect them to a centralized login page and collect their username and password there. This allows your authorization server to determine whether they have already authenticated and skip the login form entirely when it is not needed.
```

Reviewer: Just removed the "see here for more information" as it is not needed, the note can have the link.

::: note

> Collecting credentials only at the centralized login page will reduce the surface area for a potential leak of user secrets. It will also reduce the need to collect credentials unnecessarily.

Reviewer: ```suggestion
Collecting credentials only at the centralized login page will reduce the surface area for a potential leak of user secrets. It will also reduce the need to collect credentials unnecessarily.  See [Universal Login](#universal-login) for more information.
```

:::

## Anomaly Detection

> The reason that authentication systems are important is to prevent bad actors from accessing applications and user data that they should not. We want to place as many barriers in between those bad actors and access to our system as possible. One of the easiest ways to do this is to ensure that your anomaly detection with Auth0 is configured correctly. Take a moment to [read the docs on anomaly detection](/anomaly-detection) and ensure it is working correctly for you behind the scenes.

Reviewer 2: ```suggestion
The reason that authentication systems are important is to prevent bad actors from accessing applications and user data that they should not. We want to place as many barriers as possible between those bad actors and access to our systems. One of the easiest ways to do this is to ensure that your anomaly detection with Auth0 is configured correctly, so take a moment to [read the docs on anomaly detection](/anomaly-detection) and ensure that it's working correctly for you.
```

::: note

> Anomaly Detection is handled by Auth0 and provides a great security feature for your product. If you are going to utilize it, ensure that you have set up your SMTP provider and configured your email templates before turning on email delivery to your users.

Reviewer 2: ```suggestion
Anomaly Detection is handled behind the scenes by Auth0 and provides a great security feature for your product. If you're going to utilize it, ensure that you have set up your SMTP provider and configured your email templates before turning on email delivery to your users.
```

:::

## Application Integration

> Once you have figured out how you want to authenticate your users, the next step is to determine how and when you will initiate that authentication. Each application may have its own starting point.

Reviewer 2: ```suggestion
Once you've figured out how you want to authenticate your users, the next step is to determine how and when you will initiate that authentication. Each application will typically have its own starting point.
```

Reviewer 2: I favour contraction when writing in a conversational style - I find it makes the writing less "formal". But that's just personal preference :-)

Reviewer: ```suggestion
Once you've figured out how you want to authenticate your users, the next step is to determine how and when you will initiate that authentication. Each application should initiate authentication itself; how and when it initiates authentication will be particular to that application.  There are common patterns, however, that can help make it easier.
```

Reviewer: @PeterGFernandez I added some more color here as I think it is important to state that each application should be talking to the authentication server directly instead of relying on another application to tell it about the user.

::: warning

> Native Mobile Applications (and desktop applications) should use the system browser for authentication, or they open themselves up to additional security risks.

Reviewer 2: Seems to be missing the link to the recommended guidance concerning this

Reviewer: ```suggestion
Native Mobile Applications (and desktop applications) should use the system browser for authentication, or they open themselves up to additional security risks. See [Native vs Browser Login on Mobile](https://auth0.com/docs/design/browser-based-vs-native-experience-on-mobile).
```

:::

> As discussed, OpenID Connect (OIDC) is the most frequently used industry standard protocol when it comes to customer facing applications. Figuring out which OIDC flow to use is your first task, and you will want to start by reviewing guidance concerning the various grant types here.  

Reviewer 2: Probably worthwhile adding a link to where OIDC is discussed as the most frequently used industry standard protocol when it comes to B2C.

Reviewer: My comment there was more related to our experience with Auth0 customers and not an industry wide data point.  Perhaps it can just be reworded:
```suggestion
As discussed, we have found that most Auth0 customers use OpenID Connect (OIDC) as their industry standard protocol when it comes to customer facing applications. Figuring out which OIDC flow to use is your first task, and you will want to start by reviewing guidance concerning the various grant types here.  
```

Reviewer 2: Sounds good, though I was more referring to the "As discussed" bit at the start of the paragraph. It would probably be worthwhile adding a link to the "where" it was discussed in the first place :-)

Author: Added link to grant type mapping

> If you want to allow anonymous users access to any part of our application then you need to determine if you will be redirecting right away or prompting your users to redirect only when required (or perhaps some combination of both; see here for further discussion). If users can deep link to a protected version (or area) of your site then you will need to determine the links to your application that will result in an automatic redirect to Auth0 (see here for further information). 

Reviewer 2: ```suggestion
If you want to allow anonymous user access to any part of our application then you need to determine at which point you'll need (or want) to prompt your users to redirect. If users can deep link to a protected version (or area) of your site then you will need to determine the links to your application that will result in an automatic redirect to Auth0 (see here for further information). 
```

Reviewer: Missing the link there too.

### Anonymous Access

> It is important to consider the user experience when someone first comes to the application. If your application supports anonymous user access (this is quite common for eCommerce applications). There are different scenarios within this context:

Reviewer 2: ```suggestion
It is important to consider the user experience when someone first comes to the application. If your application supports anonymous user access (quite common for eCommerce applications) there are different scenarios to consider:
```

* They are returning to the application after having already logged in, or
* This is the first time they are accessing the application:
    * They have already accessed a different application that uses the same Auth0 tenant,
    * They have never (or not in a long time) authenticated on this device or browser.

> When an anonymous user accesses your application, it can often be desirable for the application to discover if the user has already logged into a different application in the same family, or to remember this user even if the application is a SPA with no state. For example, if you can determine that the user is already logged in, you might decide to have the header of the application skip displaying a login button and instead have an account or profile menu for the user. To accomplish this you will want to utilize "silent authentication". Silent authentication will allow you to check to see if the user is logged in without prompting them to log in if they are not. Then the application can present a login button if necessary. If the user is logged in already, however, then you will receive tokens and will not have to present the user with a login button again.

Reviewer 2: ```suggestion
When an anonymous user accesses your application it can often be desirable for the application to discover if the user has already logged into a different application in the same family, or to remember this user even if the application is a SPA with no state. For example, if you can determine that the user is already logged in you might decide to have the header of the application skip displaying a login button and instead have an account or profile menu for the user. To accomplish this you will want to utilize "silent authentication". Silent authentication will allow you to check to see if the user is logged in without prompting them to log in if they are not. Then the application can present a login button if necessary. If the user is logged in already, however, then you will receive tokens and will not have to present the user with a login button again.
```

::: note
Checking for a login session by redirecting to Auth0 can be really helpful for your application, but if this will result in a lot of requests it is important to employ some sort of throttling mechanism to avoid latency and/or rate limiting.
:::

### Deep linking to protected endpoints

> There are a variety of reasons why someone might link directly to a particular page within your application that is only accessible by authenticated users. If this is possible for your application you should automatically redirect your user to Auth0 if they are not authenticated. Once they authenticate and the authorization server returns them to your application, you can [redirect them](/users/guides/redirect-users-after-login) to where they intended to go in the first place.

Reviewer: If we end up with a resources section, this would be a good example of something that could move to it.

::: note
Most modern authentication frameworks support execution of middleware for redirecting to an authorization server such as Auth0. Ensure yours is configurable, can check expirations, and (for confidential clients) supports refresh tokens.
:::

### Authenticating the user

> Authentication is the process of verifying who a user is (that they are who they say they are). The result of authentication in this context is an ID Token. This token contains information about the user and should only be able to be obtained if the user authenticates using one or multiple factors as defined by the authorization server (the most common form being user id and password as discussed here). There are a few things you may also need to consider in addition to obtaining an ID Token:

Reviewer 2: ```suggestion
The result of authentication in an OIDC context is an ID Token. This token contains information about the user and should only be able to be obtained if the user authenticates using one or more factors as defined by the authorization server (the most common form being user id and password as discussed here). There are a few things you may also need to consider in addition to obtaining an ID Token:
```

* Do we also need an Access Token in order to call a shared API?
* Is our application a Native application (mobile or desktop) or do I need a Refresh Token?
    * If Yes, for more information see Authorization Code Grant with PKCE

>     * If No, for more information see Implicit Grant

Reviewer: Need links for "for more information"

Author: Added links

::: warning
Before you go live, you should ensure that only the grants that you are using for each application [are enabled in your configuration for the client](/applications/guides/update-grant-types-dashboard).
:::

### Implicit Grant

> If all your application needs is the ID Token and you are using a browser-based application, then you can always use the Implicit Grant to get your ID Token. This is a simple authentication flow and should be supported by your SDK (depending on the language you are developing in).

Reviewer 2: ```suggestion
If all your application needs is the ID Token and the application is browser-based, then you can always use the Implicit Grant to get your ID Token. This is a simple authentication flow and should be supported by your SDK (depending on the language you are developing in).
```

> If you need a refresh token so that you can re-check whether any of the information about the user has changed, then you must use the [authorization code grant](/api-auth/tutorials/authorization-code-grant).

Reviewer 2: ```suggestion
If you need a [Refresh Token](/tokens/refresh-token/current) so that you can obtain a new Access Token or ID Token without having to re-authenticate the user, then you must use the [authorization code grant](/api-auth/tutorials/authorization-code-grant).
```
